File: Data Structures & Algorithms/subsets/submission-4.cpp
class Solution {
public:

    void DFS(vector<int>&nums,int indx, vector<int>sofar, vector<vector<int>>&res)
    {
        res.push_back(sofar);
        for(int i = indx+1;i<nums.size();i++)
        {
            sofar.push_back(nums[i]);
            DFS(nums,i,sofar,res);
            sofar.pop_back();
        }

    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>>res;
        DFS(nums,-1,{},res);
        return res;
        
    }
};

File: Data Structures & Algorithms/n-th-tribonacci-number/submission-4.py
class Solution:
    def tribonacci(self, n: int) -> int:
        cache = {}
        def dfs(x):
            if x == 0:
                return 0
            if x < 3:
                return 1
            if x not in cache:
                cache[x] = dfs(x-1)+dfs(x-2)+dfs(x-3)
            return cache[x]
        return dfs(n)
            
        